File: services/auditd/testing/production/evidence/map-placement-target.ts
import type {
  PlacementAuditTargetEvidence
} from "../audit-event-evidence.js";

export function mapPlacementTarget(
  kind: number,
  tenantId: string | null,
  workspaceId: string | null,
  accountPrincipalId: string | null
): PlacementAuditTargetEvidence {
  switch (kind) {
    case 1:
      return { kind: "global" };
    case 2:
      return {
        kind: "tenant",
        tenantId: requireValue(tenantId)
      };
    case 3:
      return {
        kind: "workspace",
        tenantId: requireValue(tenantId),
        workspaceId: requireValue(workspaceId)
      };
    case 4:
      return {
        kind: "user",
        tenantId: requireValue(tenantId),
        accountPrincipalId: requireValue(accountPrincipalId)
      };
    default:
      throw new Error("Stored audit target kind is invalid");
  }
}

function requireValue(value: string | null): string {
  if (value === null) {
    throw new Error("Stored audit target is incomplete");
  }
  return value;
}

File: services/auditd/csharp/src/CtlFlow.Audit.Auditd.Service/Grpc/Requests/MapPlacementTarget.cs
using CtlFlow.Audit.Auditd.Domain.Principals;
using CtlFlow.Audit.Auditd.Domain.Tenants;
using CtlFlow.Audit.Auditd.Domain.Workspaces;
using DomainTarget =
    CtlFlow.Audit.Auditd.Domain.Events.PlacementAuditTarget;
using DomainTargetKind =
    CtlFlow.Audit.Auditd.Domain.Events.PlacementTargetKind;
using WireTarget = CtlFlow.Audit.V1.PlacementAuditTarget;

namespace CtlFlow.Audit.Auditd.Service.Grpc.Requests;

internal static partial class AuditRequests
{
    private static async ValueTask<DomainTarget> MapPlacementTarget(
        WireTarget? value,
        CancellationToken cancellation)
    {
        if (value is null)
        {
            throw new ArgumentException("Placement target is required");
        }

        return value.TargetCase switch
        {
            WireTarget.TargetOneofCase.Global => new DomainTarget(
                DomainTargetKind.Global,
                null,
                null,
                null),
            WireTarget.TargetOneofCase.Tenant => new DomainTarget(
                DomainTargetKind.Tenant,
                await TenantId.Parse(
                    value.Tenant.TenantId,
                    cancellation),
                null,
                null),
            WireTarget.TargetOneofCase.Workspace => new DomainTarget(
                DomainTargetKind.Workspace,
                await TenantId.Parse(
                    value.Workspace.TenantId,
                    cancellation),
                await WorkspaceId.Parse(
                    value.Workspace.WorkspaceId,
                    cancellation),
                null),
            WireTarget.TargetOneofCase.User => new DomainTarget(
                DomainTargetKind.User,
                await TenantId.Parse(
                    value.User.TenantId,
                    cancellation),
                null,
                await AccountId.Parse(
                    value.User.AccountPrincipalId,
                    cancellation)),
            _ => throw new ArgumentException(
                "Placement target is required")
        };
    }
}
